Mark phase prefetching.

PeterSolMS opened the pull request:
This adds prefetching to the mark phase.

The idea is that once we have established that an object is in one of the generations we want to collect, we prefetch its memory before we determine whether we have marked it already. This is because the mark bit is in the object itself, and thus requires accessing the object's memory.

As the prefetching will take some time to take effect, we park the object in a queue (see type mark_queue_t below). We then retrieve an older object from the queue, and test whether it has been marked. This should be faster, because we have issued a prefetch for this older object's memory a while back.

In quite a few places we now need to drain the queue to ensure correctness - see calls to drain_mark_queue().

PeterSolMS committed:
Mark phase prefetching.

PeterSolMS committed:
Fix Linux arm64 build issue with __prefetch.

adityamandaleeka commented on src/coreclr/gc/gc.cpp:
```diff
@@ -23606,6 +23610,114 @@ BOOL ref_p (uint8_t* r)
+
+// retrieve a newly marked object from the queue
+// returns nullptr if there is no such object
+uint8_t* mark_queue_t::drain()
```
nit: Should this be renamed to something that better implies that it just marks one object and returns it (maybe `markNext` or something)? `drain` implies completely emptying it out IMO.

Maoni0 replied:
I think of this method as draining but just needs to return if there's still objects to mark. it does drain the `slot_table` at the end when all slots become null (and that's the end goal, to have all slots become null).

PeterSolMS replied:
How about `get_next_marked`, slight variation on Aditya's suggestion?

Maoni0 replied:
then I would probably do `get_next_to_mark` since you are getting an object to do the mark work on.

PeterSolMS replied:
Well, the object returned is already marked, so `get_next_to_mark` doesn't seem entirely right either. Can't think of anything better than `get_next_marked`.

jkotas commented on src/coreclr/gc/gc.cpp:
```diff
@@ -23543,14 +23540,21 @@ void gc_heap::save_post_plug_info (uint8_t* last_pinned_plug, uint8_t* last_obje
+#endif
+    _mm_prefetch((const char*)addr, _MM_HINT_T0);
+#elif defined(TARGET_ARM64) && defined(TARGET_WINDOWS)
+    __prefetch((const char*)addr);
```
`__builtin_prefetch` should work on non-Windows

jkotas replied:
https://clang.llvm.org/docs/LanguageExtensions.html describes the arguments.

Maoni0 replied:
ah right, on linux we should use `__buildin_prefetch`.

PeterSolMS replied:
Thanks Jan - it looks like calling it with the default arguments should be just fine for our purposes.

Maoni0 commented:
some results I measured on a 1st party prod workload, I had 3 machines - 
+ baseline is without the change and 2 machines (prefetch0 and prefetch1) were with the change. 
+ the workload runs with Server GC 48 heaps, samples/mb is # of CPU samples in mark_phase to promote an mb so lower is better.

<html>
<body>
<!--StartFragment--><!DOCTYPE html>

gen0  GCs |   | samples/mb | diff  %
-- | -- | -- | --
run0 | baseline0 | 5.49 |  
  | prefetch0 | 5.29 | -3.64%
  | prefetch1 | 5.21 | -5.10%
run1 | baseline1 | 5.75 |  
  | prefetch0 | 5.47 | -4.87%
  | prefetch1 | 5.52 | -4.00%
  |   |   |  
gen1  GCs |   | samples/mb | diff  %
run0 | baseline0 | 3.46 |  
  | prefetch0 | 3.3 | -4.62%
  | prefetch1 | 3.25 | -6.07%
run1 | baseline1 | 3.6 |  
  | prefetch0 | 3.35 | -6.94%
  | prefetch1 | 3.39 | -5.83%

<!--EndFragment-->
</body>
</html>

PeterSolMS committed:
Rework implementation of Prefetch on the different architectures/operating systems.

Checkin tests showed issue traced to missing drain_mark_queue() call in WKS version of scan_dependent_handles.

PeterSolMS committed:
Adress code review feedback: rename mark_queue_t::drain to mark_queue_t::get_next_marked.

PeterSolMS committed:
Merge with main.
